src/prop.c

```diff
@@ -117,12 +147,148 @@ SEXP prop_(SEXP object, SEXP name) {
+    if (TYPEOF(no_recurse_list) == LISTSXP) {
+      // if there is a 'no_recurse' list, then this is not the top-most prop<-
+      // call for this object, i.e, we're currently evaluating a `prop<-` call
+      // called from within a custom property setter. We should only call
+      // validate(object) once from the top-most prop<- call, after the last
+      // custom setter() has returned.
+      *should_validate_obj = FALSE;
+      if (pairlist_contains(no_recurse_list, name_sym))
+        return FALSE;
+    }
+
+    if (TYPEOF(setter) != CLOSXP)
+      return FALSE; // setter not callable
+
+    Rf_setAttrib(object, sym_dot_setting_prop,
+                 Rf_cons(name_sym, no_recurse_list));
+    return TRUE; // setter now marked non-recursive, safe to call
+
+  // optimization opportunity: combine the actions of getAttrib()/setAttrib()
+  // into one loop, so we can avoid iterating over ATTRIB(object) twice.
+}
+
+static inline
+void setter_no_recurse_clear(SEXP object, SEXP name_sym) {
+  SEXP list = Rf_getAttrib(object, sym_dot_setting_prop);
+  list = pairlist_remove(list, name_sym);
+  Rf_setAttrib(object, sym_dot_setting_prop, list);
+
+  // optimization opportunity: same as setter_callable_no_recurse
+}
+
+static inline
+void prop_validate(SEXP property, SEXP value, SEXP object) {
+
+  static SEXP prop_validate = NULL;
+  if (prop_validate == NULL)
+    prop_validate = Rf_findVarInFrame(ns_S7, Rf_install("prop_validate"));
+
+  SEXP errmsg = Rf_eval(Rf_lang4(prop_validate, property, value, object), ns_S7);
+  if (errmsg != R_NilValue) signal_error(errmsg);
+}
+
+static inline
+void obj_validate(SEXP object) {
+  static SEXP validate = NULL;
+  if (validate == NULL)
+    validate = Rf_findVarInFrame(ns_S7, Rf_install("validate"));
+
+  Rf_eval(Rf_lang4(validate, object,
+                   /* recursive = */ Rf_ScalarLogical(TRUE),
+                   /* properties = */ Rf_ScalarLogical(FALSE)),
+                   ns_S7);
+}
+
+SEXP prop_set_(SEXP object, SEXP name, SEXP check_sexp, SEXP value) {
+
+  check_is_S7(object);
+
+  SEXP name_rchar = STRING_ELT(name, 0);
+  const char *name_char = CHAR(name_rchar);
+  SEXP name_sym = Rf_installTrChar(name_rchar);
+
+  Rboolean check = Rf_asLogical(check_sexp);
+  Rboolean should_validate_obj = check;
+  Rboolean should_validate_prop = check;
+
+  SEXP S7_class = Rf_getAttrib(object, sym_S7_class);
+  SEXP properties = Rf_getAttrib(S7_class, sym_properties);
+  SEXP property = extract_name(properties, name_char);
+
+  if (property == R_NilValue)
+    signal_prop_error_unknown(object, name);
+
+  SEXP setter = extract_name(property, "setter");
+  SEXP getter = extract_name(property, "getter");
+
+  if (getter != R_NilValue && setter == R_NilValue)
+    signal_prop_error("Can't set read-only property %s@%s", object, name);
+
+  PROTECT_INDEX ipx;
+  object = Rf_shallow_duplicate(object);
+  PROTECT_WITH_INDEX(object, &ipx);
```

Comment:
```suggestion
  PROTECT_INDEX object_pi;
  object = Rf_shallow_duplicate(object);
  PROTECT_WITH_INDEX(object, &object_pi);
```

In vctrs we typically use this naming pattern to tie the index to the object a little bit